### PR TITLE
Add MapMakerRing class

### PR DESCRIPTION
**Pull request by AtreyeeS**
This adds a `ImageMaker` for analysis using ring background

### CODE REVIEW

**Comment by registerrier** [comment]
Hi @AtreyeeS, could you rebase?

**Comment by AtreyeeS** [comment]
Hello! This PR 

- Refactors `MapMaker` and `MapMakerObs` so that the field of view mask is computed inside `MapMakerObs

- Changes the image creation in `MapMaker` a bit: `MapMaker.run()` will run `MapMaker` for all observations whereas `MapMaker.run_images()` will make 2D images. This replaces the `make_images()`

- Introduces `MapMakerRing` for use with RingBackgroundEstimator.

I think that there might be a mistake in the computation of the significance map. The `exposure_on` must be correlated with a TopHat kernel, but what about the other convolutions. @registerrier  and @adonath : please comment.

File: gammapy/cube/make.py [hunk 1]
```diff
@@ -257,3 +297,359 @@ def _check_selection(selection):
...
+    background_estimator : `~gammapy.background.RingBackgroundEstimator`
+        or `~gammapy.background.AdaptiveRingBackgroundEstimator`
+        Ring background estimator or something with an equivalent API.
+
+    """
+
+    def __init__(
+        self, geom, offset_max, exclusion_mask=None, background_estimator=None
+    ):
+        super(MapMakerRing, self).__init__(
```
Comment (adonath) [inline gammapy/cube/make.py]: In Python 3 `super` can be used without passing `self`. Just write `super().__init__`

File: gammapy/cube/make.py [hunk 2]
```diff
@@ -257,3 +297,359 @@ def _check_selection(selection):
...
+            geom=geom,
+            offset_max=offset_max,
+            exclusion_mask=exclusion_mask,
+            geom_true=None,
+        )
+        self.background_estimator = background_estimator
+
+    def _stack(self, maps_obs, obs_maker):
```
Comment (adonath) [inline gammapy/cube/make.py]: Would it make sense to re-use this is in `MapMaker` as well?

File: gammapy/cube/make.py [hunk 3]
```diff
@@ -257,3 +297,359 @@ def _check_selection(selection):
...
+        selection = ["counts", "exposure", "background"]
+        maps = self._get_empty_maps(selection)
```
Comment (adonath) [inline gammapy/cube/make.py]: This I don't understand. Why does `_stack()` always creates a new empty total map? Does the stacking work at all in this case?

File: gammapy/cube/make.py [hunk 4]
```diff
@@ -257,3 +297,359 @@ def _check_selection(selection):
...
+            exposure: The uncorrelated on exposure map
+            exposure_off: The off exposure map convolved with the ring
+            alpha: The alpha map (1/exposure map)
+            off: The off map
+            background_ring: The ring background map ( = alpha * off)
+            exclusion: The exclusion mask
+
+        """
+
+        map_list = []
```
Comment (adonath) [inline gammapy/cube/make.py]: This is not clear to me (also see my comment above). Does` _run()` return the maps per observations or the stacked ones?

File: gammapy/cube/make.py [hunk 5]
```diff
@@ -257,3 +297,359 @@ def _check_selection(selection):
...
+            exposure: The uncorrelated on exposure map
+            exposure_off: The off exposure map convolved with the ring
+            alpha: The alpha map (1/exposure map)
+            off: The off map
+            background_ring: The ring background map ( = alpha * off)
+
+        """
+        return self._run(observations, sum_over_axis=False)
+
+    def significance_map(self, maplist, convolution_radius=None):
```
Comment (adonath) [inline gammapy/cube/make.py]: Maybe reuse [compute_lima_on_off_image](https://docs.gammapy.org/0.10/api/gammapy.detect.compute_lima_on_off_image.html#gammapy.detect.compute_lima_on_off_image) here?

File: gammapy/cube/make.py [hunk 6]
```diff
@@ -257,3 +297,359 @@ def _check_selection(selection):
...
+            theta = 1
+        tophat = Tophat2DKernel(theta)
+        tophat.normalize("peak")
+        selection = ["on", "off", "exposure_on", "exposure_off"]
+
+        stacked_map = {}
+        for name in selection:
+            stacked_map[name] = Map.from_geom(maplist[0]["counts"].geom)
+
+        for amap in maplist:
```
Comment (adonath) [inline gammapy/cube/make.py]: OK, I see, the stacking is implemented here. But how can the user just get a stacked counts or background map?

File: gammapy/cube/make.py [hunk 7]
```diff
@@ -257,3 +297,359 @@ def _check_selection(selection):
...
+        Parameters
+        ----------
+        maplist : list of dictionaries of maps
+
+        idx : tuple, optional
+            Tuple of scalar indices for each non spatial dimension of the map.
+            Tuple should be ordered as (I_0, ..., I_n).
+
+        """
+        excess_map = self.excess_map(maplist)
```
Comment (adonath) [inline gammapy/cube/make.py]: This returns a stacked excess map, but what if the user wants to plot `excess` and `significance` only for a single observation?

File: gammapy/cube/make.py [hunk 8]
```diff
@@ -257,3 +297,359 @@ def _check_selection(selection):
...
+
+        """
+
+        significance_map = self.significance_map(maplist, convolution_radius)
+        if significance_map.geom.is_image:
+            return significance_map * self.exclusion_mask.get_image_by_idx([0])
+        else:
+            return significance_map * self.exclusion_mask
+
+    def excess_map(self, maplist):
```
Comment (adonath) [inline gammapy/cube/make.py]: This method does not use `self`, in this case it should be declared as as `@staticmethod`.

**Comment by lmohrmann** [comment]
Hi @AtreyeeS,
can you let me know on what time scale this PR will be finished? As I told you, I have a minor change that I would like to make in `cube/make.py`. Would it be OK if I open a new PR for this and you resolve conflicts later (should not be difficult...)?

**Comment by AtreyeeS** [comment]
Yes, please go ahead! I can rebase if necessary

**Pull request merged**